### PR TITLE
New FPGA Tutorial mem_channel

### PR DESCRIPTION
**Pull request by mohammadfawaz**
## Description

Creating a new tutorial that describes the usage of the `mem_channel` property and the `-Xsno-interleaving` flag.

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+The Intel® oneAPI DPC++ compiler allows you to avoid this area overhead by
+disabling burst-interleaving and assigning buffers to invidual channels. There
+are two advantages to such configuration:
+1. A simpler global memory interconnect is built which requires a smaller
+   amount of FPGA resources than the interconnect needed for the
+   burst-interleaving configuration.
+2. Potential improvements to the global memory bandwidth utilization due to
+   less contention at each memory channel.
+
```
Comment (mohammadfawaz) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Add a sentence saying that using the flag without the property or vice versa can lead to functional incorrectness. 

Edit. That's actually not the case.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+This FPGA tutorial demonstrates an example of using the `mem_channel` buffer
+property in conjuction with the `-Xsno-interleaving` flag to reduce the amount
+of resources required to implement a DPC++ FPGA design.
+
+By default, the Intel® oneAPI DPC++ compiler configures each global memory type
+in a burst-interleaved manner where memory words are interleaved across the
+available memory channels. This usually leads to better throughput because it
+prevents load imbalance by ensuring that memory accesses do not favor one
+external memory channel over another. Moreover, burst-interleaving enables you
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Maybe add a link to our documentation that talks about burst interleaving in more detail. I was hoping to see a diagram of it in action, but no need to add one here if there is one in our official docs.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+   burst-interleaving configuration.
+2. Potential improvements to the global memory bandwidth utilization due to
+   less contention at each memory channel.
+
+Burst-interleaving should only be disabled in situations where satisfactory
+load balancing can be achived by assigning buffers to individual channels.
+Otherwise, the global memory bandwidth utilization may be reduced which will
+negatively impact the throughput of your design. 
+
+To disable burst-interleaving, you need to assign a memory channel to each
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: What does this mean? The act of using mem_channel disables burst-interleaving? Why do we need the flag then?

Maybe reword if that is not what you intended to communicate

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+buffer using the `mem_channel` buffer property:
+```c++
+buffer a_buf(a_vec, {property::buffer::mem_channel{1}});
+buffer b_buf(b_vec, {property::buffer::mem_channel{2}});
+```
+The ID of the lowest available memory channel is 1. You also need to pass the
+`-Xsno-interleaving` flag to your `dpcpp` command.
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: This needs to be more clear with the sentence above

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+Running `./mem_channels_interleaving.fpga`:
+```
+Vector size: 100000 
+Verification PASSED
+
+Kernel execution time: 0.004004 seconds
+Kernel throughput 749.230914 MB/s
+```
+
+Running `./mem_channels_interleaving.fpga`:
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Can we name the two executables something different to avoid confusion which one is with interleaving and which one is without.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+constexpr double kNs = 1e9;             // number of nanoseconds in a second
+
+// Forward declare the kernel name in the global scope.
+// This FPGA best practice reduces name mangling in the optimization reports.
+class VecAdd;
+
+event runVecAdd(sycl::queue &q, const std::vector<int> &a_vec,
+                const std::vector<int> &b_vec, const std::vector<int> &c_vec,
+                std::vector<int> &sum_vec) {
+#if defined(NO_INTERLEAVING)
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp]: Nothing to do with your tutorial, but I dislike how C++ doesn't have a alternate way of removing code at compile time like this instead of this pre-processor directive.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        sum[i] = a[i] + b[i] + c[i];
+    });
+  });
+  return e;
+}
+
+// Initialize the vector from 0 to vector_size - 1
+void initialize_vector(std::vector<int> &a) {
+  for (size_t i = 0; i < a.size(); i++)
+    a.at(i) = i;
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp]: consider using this 
https://en.cppreference.com/w/cpp/algorithm/iota

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  ulong t1_kernel, t2_kernel;
+  double time_kernel;
+
+  // Create input and output vectors
+  std::vector<int> a, b, c, sum_fpga, sum_cpu;
+  a.resize(vector_size);
+  b.resize(vector_size);
+  c.resize(vector_size);
+  sum_fpga.resize(vector_size);
+  sum_cpu.resize(vector_size);
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp]: pro move with the resize. I love it

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 1 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md:54]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 3 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 5 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed. This was a typo actually - thanks for catching it.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp (hunk 7 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/mem_channels.cpp]: Done.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 4 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 2 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+To disable burst-interleaving, you need to pass the `-Xsno-interleaving` flag
+to your `dpcpp` command. With interleaving disabled, you can now specify, using
+the `mem_channel` property, in which memory channel each buffer should be
+allocated:
+```c++
+buffer a_buf(a_vec, {property::buffer::mem_channel{1}});
+buffer b_buf(b_vec, {property::buffer::mem_channel{2}});
+```
+The channels are assumed to have IDs `1,2,...,N` where `N` is the number of
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Why do you say we're assuming this, instead of knowing it?  Where can this assumption be verified?  Could this be different for different BSPs?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+available channels.
+
+For FPGA boards that have multiple memory types, it is possible to select which
+memory you want to disable burst-interleaving for by passing the memory type to
+the `-Xsno-interleaving` flag: `-Xsno-interleaving=<global_memory_type>`. The
+memory type is usually indicated in the board specification XML file.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: 'usually'?  If it's not in the XML file, then where is it?

Should we add a note about where to find the board specification XML file?  Something like 'The board specification XML file (boardspec.xml <or whatever the actual name is> can be found in the root directory of your BSP.  For example, for the PAC A10 board, you can find it . . blah blah"

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+
+An important observation is that the `mem_channel` property is a runtime
+specification which means you can change the selected channel ID for each
+buffer without having to recompile the device code. In other words, when you
+change the channel IDs, you only need to recompile the host code since
+`mem_channel` has no impact on the RTL generated by the compiler.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Add a reference to the fast_recompile tutorial here, as a reference for how to recompile only host code?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+
+### Tutorial Design 
+The basic function performed by the tutorial kernel is an addition of 3
+vectors. When burst-interleaving is disabled, each buffer is assigned to a
+specific memory channel depending on how many channels are available. 
+
+Due to the nature of this design and the fact that all the buffers have the
+same size, we expect accesses to the different memory channels to be well
+balanced. Therefore, we can predict that disabling burst-interleaving will not
+impact the throughput of the design, and so, it is likely beneficial to disable
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Delete the comma after 'and so'

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+interleaving to avoid the area overhead imposed by the interleaving logic.
+
+This tutorial requires compiling the source code twice: once with the
+`-Xsno-interleaving` flag and once without it. In the `CMakeLists.txt` file,
+the macro `NO_INTERLEAVING` is defined when the `-Xsno-interleaving` flag is
+passed to the `dpcpp` command. The macro controls whether the buffers are
+created with our without the `mem_channel` property.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: What happens if my code uses the mem_channel property, but the no-interleaving flag is not specified when compiling?  Whatever the answer, should we mention it here?

Corollary to that question, what happens if we compile with no-interleaving but don't use the mem_channel property?  Will all buffers end up in the same memory?  Or is the compiler free to choose any buffer for each memory?  In cases where you don't care about memory bandwidth at all that might be fine.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/sample.json [hunk 14]
```diff
@@ -0,0 +1,55 @@
+{
+  "guid": "8EC73D58-B010-4080-9785-B5D9DB13EF3F",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/sample.json]: Double checking that you remembered to generate a unique GUID?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+set(FPGA_EARLY_IMAGE_NO_INTERLEAVING ${TARGET_NAME}_no_interleaving_report.a)
+# The compile output is not an executable, but an intermediate compilation result unique to DPC++.
+add_executable(${FPGA_EARLY_IMAGE_INTERLEAVING} ${SOURCE_FILE})
+add_executable(${FPGA_EARLY_IMAGE_NO_INTERLEAVING} ${SOURCE_FILE})
+add_custom_target(report DEPENDS ${FPGA_EARLY_IMAGE_INTERLEAVING} ${FPGA_EARLY_IMAGE_NO_INTERLEAVING})
+set_target_properties(${FPGA_EARLY_IMAGE_INTERLEAVING} PROPERTIES COMPILE_FLAGS "${HARDWARE_COMPILE_FLAGS}")
+set_target_properties(${FPGA_EARLY_IMAGE_INTERLEAVING} PROPERTIES LINK_FLAGS "${HARDWARE_LINK_FLAGS} -fsycl-link=early")
+
+if (FPGA_BOARD STREQUAL "intel_a10gx_pac:pac_a10")
+  set_target_properties(${FPGA_EARLY_IMAGE_NO_INTERLEAVING} PROPERTIES COMPILE_FLAGS "${HARDWARE_COMPILE_FLAGS} -DNO_INTERLEAVING")
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/CMakeLists.txt]: What happens when we compile this design to some non-PAC BSP?  You've assumed everything other than pac_a10 will have four channels.  What happens when we compile to our a10_ref or s10_ref BSPs, which only have one channel?  Will we get errors?  What about agilex_ref?  I'm not sure if it has 2 or 4 channels.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 9 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 10 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md (hunk 13 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/README.md]: Addressed.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/CMakeLists.txt (hunk 15 above)
Comment (mohammadfawaz) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/CMakeLists.txt]: Here's what I did to address this:
* Added  two macros: `TWO_CHANNELS` and `FOUR_CHANNELS`.
* Defined `TWO_CHANNELS` for PAC A10.
* Defined `FOUR_CHANNELS` for PAC S10.
* Not define anything for any other board but ask the reader to set this macro correctly for whatever customer board they're using. For
So, for a10_ref and s10_ref, `-Xsno-interleaving` won't change the behavior of the design because only 1 channels exists anyways. For Agilex, the no-interleaving version won't have any macros defined. Once we start mentioning Agilex in the tutorials, we then have to define the `FOUR_CHANNELS` macro for that target.

**Review by mdbtucker — APPROVED** [review]
Looks good to me, thanks Mohammad.  Please close this PR (don't merge!) and create a new PR to the main repo master branch.

Also, what is the status of a regtest for this new sample?  Please add me to the swarm review.
